### PR TITLE
cli: Add `onboard` command

### PR DESCRIPTION
**Pull request by marcuseide**
## Hey, I just made a Pull Request!

Adding a `onboard` command to `@backstage/cli`. The intention with this new command is to provide a somewhat opinionated way to assist, or guide, a user that just created a new Backstage App in setting it up. 

In this first iteration a user can get help setting up Github authentication - we are using the term "authentication" deliberately generic - as in hooking up auth-backend, adding oauth credentials and providing a functioning sign-in page. This is aimed towards users that might not have a deep understanding on the complexity of authentication but rather wants to see the benefits of Backstage directly and without sifting through too many docs.

Our plan is to expand this into more advanced setups and options going forward.

This should be considered work-in-progress. We've used `{hidden: true}` to indicate that this is not ready for use yet, and we intend to work iteratively on this while at the same time collecting feedback from eager adopters.

```
$ yarn backstage-cli onboard
? Do you want to set up Authentication for this project? Yes
? Please select a provider: GitHub

    To add GitHub authentication, you must create an OAuth App from the GitHub developer settings: https://github.com/settings/developers
    The Homepage URL should point to Backstage frontend, while the Authorization callback URL will point to the auth backend.

    Settings for local development:
    
      Homepage URL: http://localhost:3000
      Authorization callback URL: http://localhost:7007/api/auth/github/handler/frame

    You can find the full documentation page here: https://backstage.io/docs/auth/github/provider
    
? What is your GitHub username? marcuseide
? What is your Client Id? 123456789
? What is your Client Secret? 1234567890987654321
? Are you using GitHub Enterprise? No
Setting up GitHub Authentication for you...
  Validating    credentials ✔ 
  Updating      app-config.local.yaml ✔ 
  Creating      user-info.yaml ✔ 
  Patching      01-github.Add-SignInPage.patch ✔ 
  Patching      02-github.Use-signIn-resolver.patch ✔ 
Done setting up GitHub Authentication!
You can now start your app with yarn dev
```

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: packages/cli/src/commands/onboard/auth/patches/01-github.Add-SignInPage.patch [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
++        auto
++        provider={{
++          id: 'github-auth-provider',
++          title: 'GitHub',
++          message: 'Sign in using GitHub',
++          apiRef: githubAuthApiRef,
++        }}
++      />
++    ),
++  },
```
Comment (Rugvip) [inline packages/cli/src/commands/onboard/auth/patches/01-github.Add-SignInPage.patch:26]: Just wanna call out that this as well as the sign-in resolver are things that we want to figure out how to more or less eliminate from default code configuration, i.e. make them completely config driven instead.

No action required on this PR but just highlighting that we shouldn't invest large amounts of time into making this code configuration experience better. At that point it's better to start up the work to make this config driven instead.

File: packages/cli/src/commands/onboard/auth/github/index.ts [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+  await Task.forItem(
+    'Updating',
+    APP_CONFIG_FILE,
+    async () => await updateConfigFile(APP_CONFIG_FILE, config),
+  );
+  await Task.forItem(
+    'Creating',
+    USER_ENTITY_FILE,
+    async () => await addUserEntity(USER_ENTITY_FILE, username),
+  );
```
Comment (Rugvip) [inline packages/cli/src/commands/onboard/auth/github/index.ts:165]: Hmm, tbh I don't think we should encourage this. Managing users through YAML only makes sense for really small setups. Even as an onboarding step I think it's better to for example use a sign-in resolver that doesn't require users in the catalog.

**Review by Rugvip — COMMENTED** [review]
Nice! 👏 

Couple of initial thoughts

File: packages/cli/src/commands/onboard/auth/github/index.ts (hunk 2 above)
Comment (marcuseide) [reply packages/cli/src/commands/onboard/auth/github/index.ts]: Yeah, this was initially confusion from my part not realizing the ease of just using a custom sing-in resolver. It also came up talking to @freben about our approach earlier.
We're not really sure what we want to target here, so let me keep it in just a bit longer while we figure out what we want to do. Appreciate the input 👍

File: packages/cli/src/commands/onboard/README.md [hunk 3]
```diff
@@ -0,0 +1,23 @@
+# admin tool
+
+## Local dev
+
+```bash
+# Checkout the branch
+cd backstage
+git fetch
+git checkout goa/cli-admin-command
```
Comment (Rugvip) [inline packages/cli/src/commands/onboard/README.md:9]: Just noting that this only makes sense right now and shouldn't be merged 😅

File: packages/cli/src/commands/onboard/README.md [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+# Create a new app
+cd ..
+npx @backstage/create-app
+cd new-backstage-app
+
+# Run the modded cli in the new app
+../github/backstage/packages/cli/bin/backstage-cli admin
```
Comment (Rugvip) [inline packages/cli/src/commands/onboard/README.md:17]:
```suggestion
../github/backstage/packages/cli/bin/backstage-cli onboard
```

File: packages/cli/src/commands/onboard/README.md [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+```bash
+# Checkout the branch
+cd backstage
+git fetch
+git checkout goa/cli-admin-command
+
+# Create a new app
+cd ..
+npx @backstage/create-app
```
Comment (Rugvip) [inline packages/cli/src/commands/onboard/README.md:13]: With `--skip-install` if you wanna make it quick 😁, don't think you need deps for testing this new command right?

File: packages/cli/src/commands/onboard/auth/github/index.ts [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+              enterpriseInstanceUrl,
+            }),
+          },
+        },
+      },
+    },
+  };
+};
+
+export const github = async () => {
```
Comment (Rugvip) [inline packages/cli/src/commands/onboard/auth/github/index.ts:79]: Would prefer if we avoid this much code in the index files, rather create a separate file and re-export. Or perhaps `onboard/auth/github.ts` could be enough in this case

File: packages/cli/src/commands/onboard/auth/github/index.ts [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+  Task.log(`
+    To add GitHub authentication, you must create an OAuth App from the GitHub developer settings: ${chalk.blue(
+      'https://github.com/settings/developers',
```
Comment (Rugvip) [inline packages/cli/src/commands/onboard/auth/github/index.ts:82]: This should point to the enterprise instance if provided

File: packages/cli/src/commands/onboard/auth/patches/01-github.Add-SignInPage.patch (hunk 1 above)
Comment (Rugvip) [reply packages/cli/src/commands/onboard/auth/patches/01-github.Add-SignInPage.patch]: Actually, do you think it's worth considering moving that into scope of this effort? I do feel that a close to zero-config sign-in system would significantly simplify adoption.

**Review by Rugvip — CHANGES_REQUESTED** [review]
Some bonus comments 😁

File: packages/cli/src/commands/onboard/README.md (hunk 5 above)
Comment (tudi2d) [reply packages/cli/src/commands/onboard/README.md:13]: yeah - don't really need deps for the command. It could be annoying afterwards though, as you will probably want to run it directly after the `onboard` command & then have to wait for the install. Either way is totally fine for me! :)

File: packages/cli/src/commands/onboard/auth/patches/01-github.Add-SignInPage.patch (hunk 1 above)
Comment (mitchhentgesspotify) [reply packages/cli/src/commands/onboard/auth/patches/01-github.Add-SignInPage.patch]: > just highlighting that we shouldn't invest large amounts of time into making this code configuration experience better.

Hey, I just wanted to chip in a bit here with extra context.
My understanding is that this "onboard" command is trying to 80% reduce onboarding friction with 20% of the effort.

I absolutely agree that the _correct_ way of solving this problem is identifying each manual (or code-change) step of onboarding, then evaluating if it _should_ be manual (or config instead), or how it can otherwise be improved. However, it's a relatively high cost to change each of these:
* It's a breaking change
* We'll want to be relatively confident for each one of these changes, to avoid having to tweak them again
* There may be migration needs, etc

Fortunately, I think the "onboard" command _will_ help with improving the underlying changes: due to its experimental nature and "one-off, 'patch or bail'" approach, we have a lot of flexibility to iterate and innovate with it.
Over time, I'm expecting it to organically "smooth down" and expose a _very_ ergonomic UX to users, at which point we can learn from it to much more confidently and easily improve the underlying config/code-change requirements.

-----

In summary, I'd advocate towards deferring the work to improve the underlying config/code-change requirements for two reasons:
1. We get 80% of the UX experience without needing nearly as much planning work up-front
2. We still get to improve the underlying experience in the future, but we'll be equipped with the learnings from the "onboard" command.

Comment (Rugvip) [reply packages/cli/src/commands/onboard/auth/patches/01-github.Add-SignInPage.patch]: Yep! Just being mindful of spending too much time polishing a 💩, 😁

File: packages/cli/src/commands/onboard/config.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+const readYaml = async (file: string) => {
+  return yaml.parse(await fs.readFile(file, 'utf8'));
+};
+
+export const updateConfigFile = async <T>(file: string, config: T) => {
+  const content = fs.existsSync(file)
+    ? { ...(await readYaml(file)), ...config }
+    : { ...config };
+
+  return await fs.writeFile(
```
[the remark this answers is not in the document]
Comment (tudi2d) [reply packages/cli/src/commands/onboard/config.ts]: Thanks for the feeback! Adjusted! :)

File: packages/cli/src/commands/onboard/auth/github/index.ts (hunk 6 above)
Comment (tudi2d) [reply packages/cli/src/commands/onboard/auth/github/index.ts]: moved this (& `gitlab/index.ts`) to `providers/github.ts`

File: packages/cli/src/commands/onboard/README.md [hunk 9]
```diff
@@ -0,0 +1,22 @@
+# admin tool
```
Comment (jhaals) [inline packages/cli/src/commands/onboard/README.md:1]:
```suggestion
# onboard
```

**Review by jhaals — APPROVED** [review]
Looking good to me 👍
